### PR TITLE
ENH: Add support for numpy 2's string dtype

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Just submitting this for CI purposes, nothing to review yet.

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2821,7 +2829,7 @@ NoDefault = Literal[_NoDefault.no_default]
 
 
 def map_infer_mask(
-        ndarray[object] arr,
+        ndarray arr,
```
Comment (lithomas1) [inline pandas/_libs/lib.pyx:2832]: This will make iteration slower.

Look into ``PyArray_GETITEM``/``PyArray_ITER_NEXT`` like in ``Validator._validate``?

Maybe this can be a separate PR?

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -2869,8 +2877,8 @@ def map_infer_mask(
 @cython.boundscheck(False)
 @cython.wraparound(False)
 def _map_infer_mask(
-        ndarray[uint8_int64_object_t] out,
-        ndarray[object] arr,
+        ndarray out,
+        ndarray arr,
```
Comment (lithomas1) [inline pandas/_libs/lib.pyx:2881]: This will make iteration slower.

Look into numpy C API iteration functions.

(e.g. look at Validator._validate)

If this gives perf improvement, maybe can be split out.

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -2104,7 +2105,10 @@ def is_view(self) -> bool:
 
     @property
     def array_values(self) -> ExtensionArray:
-        return NumpyExtensionArray(self.values)
+        if self.values.dtype.kind == 'T':
+            return NumpyStringArray(self.values)
+        else:
+            return NumpyExtensionArray(self.values)
```
Comment (ngoldbaum) [inline pandas/core/internals/blocks.py:2111]: This is because NumpyExtensionArray inherit from `ObjectStringArrayMixin`, so unless I make this change there's no way to call the string method overrides in `NumpyStringArray`.

**Comment by ngoldbaum** [comment]
Closing in favor of #58394.

**Comment by ngoldbaum** [comment]
Oh wait I can't, I guess the bot will get to it if Thomas doesn't :)
